### DISCUSSION TITLE
Report functionality

### DISCUSSION THREAD

**Discussion by Manishkparmar80** [Budibase/budibase#1696] [Ideas]
E.g. i have developed form and captured input and saved record into database. Now based on saved data, i want generate report. E.g. i have booked sales order and now i want to generate sales invoice report based on save data then how to achieve the same in budibase.

**Comment by shogunpurple** [comment]
@Manishkparmar80 - thanks for raising. PDF/Report generation functionality is certainly something that is in our plans in the coming months, but we aren't sure exactly when. Watch this space though for updates.

**Comment by mjashanks** [comment]
Thanks @Manishkparmar80 . Are there any particular tools that you would normally use for this? 

An integration with another tool might be the best option for now. Would be good to get a few suggestions.

If you use a tool already, you might find [this discussion](https://github.com/Budibase/budibase/discussions/1657) useful, as a way to get your data into that tool today.

We will also be documenting our API soon, which will open it up to querying.

**Reply by KernOil** [reply]
I have been using google data studios to run my reports it's free and works great. I am having to write to an external PostgreSQL database to get the connection. I think a built-in connection to data studio would be a super quick way of putting reports into your platform.

**Reply by Manishkparmar80** [reply]
@aptkingston  Hi AndrewJRamay, i have never worked with Google data studio. Can yoy pls help me with sample screenshot on how to configure and use Google Data Studio within Budibase application and generating PDF reports directly from Budibase web application. I am using mysql external data base.

@mjashanks Hi Michael, i do not have past experiance with any of report generation tool. But may be we shall take opinion of @aptkingston as he has already worked on such tool. Also while google search i found open source tool like Carbon https://carbone.io/ who are claiming to have such functionality but i am not sure about how to configure and use it with budibase application.

**Reply by KernOil** [reply]
@Manishkparmar80 
so basically it's a free service from google it will read your data in real-time. I have TV's set up that shows reports that auto-refresh every 1 min. it's a drag and drop builder. first, you add a data source. 
![image](https://user-images.githubusercontent.com/80058238/127694059-e16621e5-d7da-41a7-884c-76fc7b45669d.png)
 

I use MySQL so I just connect directly to that DB. this is the same SQL DB that my BB app writes to.

Input your connection info 
![image](https://user-images.githubusercontent.com/80058238/127694168-4b17460c-4130-4369-a9c9-12aa714fab41.png)
 
then you can start using the builder its a bit like a dumbed-down version of a BI tool but works great for reporting you can build out templates and conventional operations.    
here is a report I built out my lab inputs data to my BB app and it will update in real-time on this report 1 of about 20 reports I have. 
![image](https://user-images.githubusercontent.com/80058238/127695151-7a5e5b54-d613-4c7b-9184-1047b2f3c1e8.png)

**Reply by KernOil** [reply]
it also has an auto email function so it will email out your reports at a given time or trigger

**Comment by Manishkparmar80** [comment]
@shogunpurple @joebudi is PDF report functionality added in July release.

Looks like using OpenPDF it might be possible but not sure how to use it in integration with Budibase. https://github.com/LibrePDF/OpenPDF

**Reply by shogunpurple** [reply]
@Manishkparmar80 no, it has not been added. Please refer to the release notes:

https://github.com/Budibase/budibase/discussions/2165

**Reply by Manishkparmar80** [reply]
@shogunpurple @joebudi so as per #2165 , even in upcoming release of Aug also PDF report functionality not considered.

**Reply by joebudi** [reply]
It was considered but we only have so much resource, so it was not chosen. We do welcome contributions from our community. 

I do agree it's an important and welcomed feature that will come soon. It's on our radar, just not in the Aug release.

**Comment by hendrikjap** [comment]
Hope you might reconsider prioritising this PDF report functionality. Happy New Year guys!

**Reply by shogunpurple** [reply]
Hi @ely2727 - it's definitely going to be on the roadmap for the next few months, it's a commonly requested one. We will update this discussion and the community in due course when we start work on it. What is your use case? What kind of features would you like to see when it comes to generating a PDF report?

Happy new year!

**Reply by hendrikjap** [reply]
You can go either no code way such as Airtable or Seatable implementation ("Page Designer") or SQL notebook way to generate report, e.g. https://evidence.dev/

Ability to recall any data to PDF report would be a differentiating factor and make Budibase stands out.

**Reply by SaschaValentin** [reply]
Hi,
some visual report builder like in Seatable where you can drag and drop the fields and compose it with static text and images would be great 😍

But maybe until this is realized you can give a hint how to integrate a js-library like https://parall.ax/products/jspdf. Or in other words: How to hand over values from forms or selected rows in Budibase to an outer JavaScript?

Another way could be integrating libraries in Budibase.

Happy New Year to all

**Reply by hendrikjap** [reply]
It's much needed improvement for budibase! Look forward to it.

**Comment by kevintosh** [comment]
I kinda hacked some solution here: [#2944](https://github.com/Budibase/budibase/discussions/2944#discussioncomment-7697418)
